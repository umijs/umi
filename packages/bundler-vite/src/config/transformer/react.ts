import reactPlugin from '@vitejs/plugin-react';
import type { IConfigProcessor } from '.';

/**
 * enable react plugin & transform umi babel to vite babel
 */
export default (function react(userConfig) {
  const config: ReturnType<IConfigProcessor> = { plugins: [] };

  // add react plugin config ability.
  if (!userConfig.vue) {
    config.plugins?.push(
      // pre-compiled rollup type is different with installed rollup type
      // so this plugin type is not compatible with config.plugins
      // @ts-ignore
      reactPlugin({
        // jsxRuntime: 'automatic',
        babel: {
          plugins: userConfig.extraBabelPlugins,
          presets: userConfig.extraBabelPresets,
        },
        ...(userConfig?.react || {}),
      }),
    );
  }

  return config;
} as IConfigProcessor);
