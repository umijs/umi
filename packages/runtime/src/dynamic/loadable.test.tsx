import { render } from '@testing-library/react';
import React from 'react';
import Loadable from './loadable';

function waitFor(delay: number) {
  return new Promise((resolve) => {
    setTimeout(resolve, delay);
  });
}

function createLoader(
  delay: number,
  loader: null | (() => void),
  error?: Error,
) {
  return () => {
    return waitFor(delay).then(() => {
      if (loader) {
        return loader();
      } else {
        throw error;
      }
    });
  };
}

function MyLoadingComponent(
  props: React.PropsWithChildren<{
    prop: string;
  }>,
) {
  return <div>MyLoadingComponent {JSON.stringify(props)}</div>;
}

function MyComponent(props: React.PropsWithChildren<{ prop: string }>) {
  return <div>MyComponent {JSON.stringify(props)}</div>;
}

afterEach(async () => {
  try {
    await Loadable.preloadAll();
  } catch (err) {}
});

test('server side rendering', async () => {
  let LoadableMyComponent = Loadable<{ prop: string }>({
    loader: createLoader(400, () => require('../__fixtures__/component')),
    loading: MyLoadingComponent,
  });

  await Loadable.preloadAll();

  let component = render(<LoadableMyComponent prop="baz" />);

  expect(component.container.innerHTML).toMatchSnapshot(); // serverside
});

test('server side rendering es6', async () => {
  let LoadableMyComponent = Loadable<{ prop: string }>({
    loader: createLoader(400, () => require('../__fixtures__/component.es6')),
    loading: MyLoadingComponent,
  });

  await Loadable.preloadAll();

  let component = render(<LoadableMyComponent prop="baz" />);

  expect(component.container.innerHTML).toMatchSnapshot(); // serverside
});

test('preload', async () => {
  let LoadableMyComponent = Loadable<{ prop: string }>({
    loader: createLoader(400, () => MyComponent),
    loading: MyLoadingComponent,
  });

  let promise = LoadableMyComponent.preload();
  await waitFor(200);

  let component1 = render(<LoadableMyComponent prop="baz" />);

  expect(component1.container.innerHTML).toMatchSnapshot(); // still loading...
  await promise;
  expect(component1.container.innerHTML).toMatchSnapshot(); // success

  let component2 = render(<LoadableMyComponent prop="baz" />);
  expect(component2.container.innerHTML).toMatchSnapshot(); // success
});

test('render', async () => {
  let LoadableMyComponent = Loadable<{ prop: string }>({
    loader: createLoader(400, () => ({ MyComponent })),
    loading: MyLoadingComponent,
    render(loaded, props) {
      return <loaded.MyComponent {...props} />;
    },
  });
  let component = render(<LoadableMyComponent prop="baz" />);
  expect(component.container.innerHTML).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component.container.innerHTML).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component.container.innerHTML).toMatchSnapshot(); // success
});

test('loadable map success', async () => {
  let LoadableMyComponent = Loadable.Map<{ prop: string }>({
    loader: {
      a: createLoader(200, () => ({ MyComponent })),
      b: createLoader(400, () => ({ MyComponent })),
    },
    loading: MyLoadingComponent,
    render(loaded, props) {
      return (
        <div>
          <loaded.a.MyComponent {...props} />
          <loaded.b.MyComponent {...props} />
        </div>
      );
    },
  });

  let component = render(<LoadableMyComponent prop="baz" />);
  expect(component.container.innerHTML).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component.container.innerHTML).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component.container.innerHTML).toMatchSnapshot(); // success
});

test('loadable map error', async () => {
  let LoadableMyComponent = Loadable.Map<{
    prop: string;
  }>({
    loader: {
      a: createLoader(200, () => ({ MyComponent })),
      b: createLoader(400, null, new Error('test error')),
    },
    loading: MyLoadingComponent,
    render(loaded, props) {
      return (
        <div>
          <loaded.a.MyComponent {...props} />
          <loaded.b.MyComponent {...props} />
        </div>
      );
    },
  });

  let component = render(<LoadableMyComponent prop="baz" />);
  expect(component.container.innerHTML).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component.container.innerHTML).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component.container.innerHTML).toMatchSnapshot(); // success
});

describe('preloadReady', () => {
  beforeEach(() => {
    (global as any).__webpack_modules__ = { 1: true, 2: true };
  });

  afterEach(() => {
    delete (global as any).__webpack_modules__;
  });

  test('undefined', async () => {
    let LoadableMyComponent = Loadable<{ prop: string }>({
      loader: createLoader(200, () => MyComponent),
      loading: MyLoadingComponent,
    });

    await Loadable.preloadReady();

    let component = render(<LoadableMyComponent prop="baz" />);

    expect(component.container.innerHTML).toMatchSnapshot();
  });

  test('one', async () => {
    let LoadableMyComponent = Loadable<{ prop: string }>({
      loader: createLoader(200, () => MyComponent),
      loading: MyLoadingComponent,
      webpack: () => [1],
    });

    await Loadable.preloadReady();

    let component = render(<LoadableMyComponent prop="baz" />);

    expect(component.container.innerHTML).toMatchSnapshot();
  });

  test('many', async () => {
    let LoadableMyComponent = Loadable<{ prop: string }>({
      loader: createLoader(200, () => MyComponent),
      loading: MyLoadingComponent,
      webpack: () => [1, 2],
    });

    await Loadable.preloadReady();

    let component = render(<LoadableMyComponent prop="baz" />);

    expect(component.container.innerHTML).toMatchSnapshot();
  });

  test('missing', async () => {
    let LoadableMyComponent = Loadable<{ prop: string }>({
      loader: createLoader(200, () => MyComponent),
      loading: MyLoadingComponent,
      webpack: () => [1, 42],
    });

    await Loadable.preloadReady();

    let component = render(<LoadableMyComponent prop="baz" />);

    expect(component.container.innerHTML).toMatchSnapshot();
  });

  test('delay with 0', () => {
    let LoadableMyComponent = Loadable<{ prop: string }>({
      loader: createLoader(300, () => MyComponent),
      loading: MyLoadingComponent,
      delay: 0,
      timeout: 200,
    });

    let loadingComponent = render(<LoadableMyComponent prop="foo" />);

    expect(loadingComponent.container.innerHTML).toMatchSnapshot(); // loading
  });
});
