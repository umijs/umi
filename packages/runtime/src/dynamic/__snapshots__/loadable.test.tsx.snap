// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadable map error 1`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":false,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`loadable map error 2`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":true,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`loadable map error 3`] = `"<div>MyLoadingComponent {\\"isLoading\\":false,\\"pastDelay\\":true,\\"timedOut\\":false,\\"error\\":{}}</div>"`;

exports[`loadable map success 1`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":false,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`loadable map success 2`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":true,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`loadable map success 3`] = `"<div><div>MyComponent {\\"prop\\":\\"baz\\"}</div><div>MyComponent {\\"prop\\":\\"baz\\"}</div></div>"`;

exports[`preload 1`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":true,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`preload 2`] = `"<div>MyComponent {\\"prop\\":\\"baz\\"}</div>"`;

exports[`preload 3`] = `"<div>MyComponent {\\"prop\\":\\"baz\\"}</div>"`;

exports[`preloadReady delay with 0 1`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":true,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`preloadReady many 1`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":false,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`preloadReady missing 1`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":false,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`preloadReady one 1`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":false,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`preloadReady undefined 1`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":false,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`render 1`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":false,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`render 2`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":true,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`render 3`] = `"<div>MyComponent {\\"prop\\":\\"baz\\"}</div>"`;

exports[`server side rendering 1`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":false,\\"timedOut\\":false,\\"error\\":null}</div>"`;

exports[`server side rendering es6 1`] = `"<div>MyLoadingComponent {\\"isLoading\\":true,\\"pastDelay\\":false,\\"timedOut\\":false,\\"error\\":null}</div>"`;
